packages/remix-react/routes.tsx

```diff
@@ -183,82 +125,56 @@ async function loadRouteModuleWithBlockingLinks(
   return routeModule;
 }
 
-function createLoader(route: EntryRoute, routeModules: RouteModules) {
-  let loader: ClientRoute["loader"] = async ({ url, signal, submission }) => {
-    if (route.hasLoader) {
-      let routeModulePromise = loadRouteModuleWithBlockingLinks(
-        route,
-        routeModules
-      );
-      try {
-        let result = await fetchData(url, route.id, signal, submission);
-
-        if (result instanceof Error) throw result;
-
-        /* TODO: Handle client side redirects
-        let redirect = await checkRedirect(result);
-        if (redirect) return redirect;
-        */
-
-        if (isCatchResponse(result)) {
-          throw new CatchValue(
-            result.status,
-            result.statusText,
-            await extractData(result)
-          );
-        }
-
-        return extractData(result);
-      } finally {
-        await routeModulePromise;
-      }
-    } else {
-      await loadRouteModuleWithBlockingLinks(route, routeModules);
-    }
-  };
-
-  return loader;
-}
-
-function createAction(route: EntryRoute, routeModules: RouteModules) {
-  let action: ClientRoute["action"] = async ({ url, signal, submission }) => {
-    let routeModulePromise = await loadRouteModuleWithBlockingLinks(
+function createDataFunction(
+  route: EntryRoute,
+  routeModules: RouteModules,
+  isAction: boolean
+): LoaderFunction | ActionFunction {
+  return async ({ request }) => {
+    let routeModulePromise = loadRouteModuleWithBlockingLinks(
       route,
       routeModules
     );
-
     try {
-      if (!route.hasAction) {
+      if (isAction && !route.hasAction) {
         console.error(
           `Route "${route.id}" does not have an action, but you are trying ` +
             `to submit to it. To fix this, please add an \`action\` function to the route`
         );
+        // TODO: Should we `return null` here like we do for loaders?  Is there
+        // a benefit to triggering a network request that will error/404?
```

Comment:
Confirm behavior here that we want to make the fetch request that we know is going to fail with a 405.  Easier for debugging?

Reply:
No need to call the fetch for the 405 here can short circuit like we do for loaders